### DISCUSSION TITLE
How to override IP adresses of outdated ProtonVPN servers?

### DISCUSSION THREAD

**Discussion by FitRiver** [qdm12/gluetun#1012] [Q&A] [answered]
The ProtonVPN servers from Austria are outdated for a while and I wasn't able to figure out how to override them in the servers.json file.

At first, I tried inserting another hostname to the servers.json file. I used the hostname "node-at-06.protonvpn.net" ProtonVPN's config file.

The log shows:

```
INFO` [storage] merging by most recent 11416 hardcoded servers and 11417 servers read from /gluetun/servers.json
INFO [storage] Using privatevpn servers from file which are 1040h0m21s more recent
```

It looks like it recognizes the newly added server and also the updated timestamp.

However, the next line states:

```
ERROR VPN settings: provider settings: server selection: the hostname specified is not valid: value is not one of the possible choices: value "node-at-06.protonvpn.net", choices available are...

(the list of all hardcoded ProtonVPN servers follows)
```

Does it mean it's not possible to add a new hostname that is not hardcoded?

I also try to re-use one of the old hostnames and only replace the IP addresses. However, it still used the old hardcoded address.

Could anyone give me a hint on how to solve this situation?

**Comment by qdm12** [comment, marked as answer]
Hello there;

So the system was rather clumsy, but you could do it by adding the server you want and then bump the timestamp to your current timestamp (find it at https://www.unixtimestamp.com/).

For example:

```json
   // ...
  "mullvad": {
    "version": 4,
    "timestamp": 1654467404,
    "servers": [
      {
        "vpn": "wireguard",
        "country": "xyz",
        "tcp": true,
        "udp": true,
        "ips": [
          "31.171.154.210"
        ]
      },
      // ...
```

HOWEVER this will be overridden once the image has a newer list of servers (higher timestamp), so one trick is to set a much bigger timestamp so it's always kept no matter what the hardcoded servers timestamp is.

NOW... because that system was stupid-ish, I just pushed 5e659dc5b391075dfc374e6e16a31ea7d64e961e which should allow you to add a field `"keep": true` to a server JSON object, and that one will be kept across updates. So ideal for manually inputted servers. I haven't tested it though, so let me know if it works or doesn't!

A bit out of scope, there is also the provider version field which you should not modify, but if the program bumps it in its hardcoded servers, your servers data will be thrown out since it means the newer gluetun program can't understand previously formed JSON (in example a field key name changed). This however is now unlikely to happen since all server models for all providers are the same model, and it would be rather troublesome to break compatibility. I'll also reset those version fields to `1` for all providers once I release a `v4.0.0` breaking-change image (not anytime soon AFAIK).

**Reply by FitRiver** [reply]
Thank you for your advice and support.

I pulled the latest image and added the `"keep": true` to the configuration.

The host uses the correct new IP address now.

Btw, should I open some issue about the outdated ProtonVPN servers, or are the configs scraped on a regular basis so they will show up in future releases of the servers.json anyway.

**Reply by qdm12** [reply]
You can batch update the servers yourself, see the Updating servers page on the github wiki. You can also set UPDATE_PERIOD to do that periodically.

Other than that, I update them in the image manually (as in automatically but I have to run the command and push the changes) usually when there is a problem with that provider. I prefer to avoid this so users get used to not being dependent on me updating hardcoded servers (i. e. if I'm on a trip or dead or whatever). I believe gluetun should log this out if you let it boot crash a few times and link to the wiki page.

**Reply by FitRiver** [reply]
Oh, I see. Thank you for the explanation. I actually read this page on the wiki multiple times as I was experimenting with the app, but I thought that the update on the "Updating servers" means only downloading the newest servers.json file from the repository and that it refreshes the IP addresses based on the hosts.

I checked that https://github.com/qdm12/gluetun/blob/master/internal/storage/servers.json (updated today) doesn't contain the newest hosts from ProtonVPN so I thought that updating manually wouldn't help.

Perhaps it would be useful to mention in the wiki that running the manual update means that Gluetun will scrape the currently valid hostnames from the providers.

**Reply by qdm12** [reply]
Done in https://github.com/qdm12/gluetun/wiki/Updating-servers#update-mechanism 😉 Thanks for the suggestion.

**Reply by FitRiver** [reply]
Thank you for updating the documentation! It's very understandable now.
